// @llamaindex/chat-ui/dist/widgets/progress-12s-B6tOi6Ka.js
'use client';
import { _ as _$1 } from '@swc/helpers/_/_extends';
import { _ } from '@swc/helpers/_/_object_without_properties_loose';
import { jsx } from 'react/jsx-runtime';
import * as ProgressPrimitive from '@radix-ui/react-progress';
import * as React from 'react';
import { c as cn } from './drawer-12s-CiogNq7V.js';

const Progress = /*#__PURE__*/ React.forwardRef((_param, ref)=>{
    var { className, value } = _param, props = _(_param, [
        "className",
        "value"
    ]);
    return /*#__PURE__*/ jsx(ProgressPrimitive.Root, _$1({
        ref: ref,
        className: cn('bg-secondary relative h-4 w-full overflow-hidden rounded-full', className)
    }, props, {
        children: /*#__PURE__*/ jsx(ProgressPrimitive.Indicator, {
            className: "bg-primary h-full w-full flex-1 transition-all",
            style: {
                transform: `translateX(-${100 - (value || 0)}%)`
            }
        })
    }));
});
Progress.displayName = ProgressPrimitive.Root.displayName;

export { Progress as P };
