// @llamaindex/chat-ui/dist/widgets/progress-12s-C5KmUjfL.cjs
'use client';
var _extends = require('@swc/helpers/_/_extends');
var _object_without_properties_loose = require('@swc/helpers/_/_object_without_properties_loose');
var jsxRuntime = require('react/jsx-runtime');
var ProgressPrimitive = require('@radix-ui/react-progress');
var React = require('react');
var drawer12s = require('./drawer-12s-ClgJ3fRT.cjs');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return n;
}

var ProgressPrimitive__namespace = /*#__PURE__*/_interopNamespace(ProgressPrimitive);
var React__namespace = /*#__PURE__*/_interopNamespace(React);

const Progress = /*#__PURE__*/ React__namespace.forwardRef((_param, ref)=>{
    var { className, value } = _param, props = _object_without_properties_loose._(_param, [
        "className",
        "value"
    ]);
    return /*#__PURE__*/ jsxRuntime.jsx(ProgressPrimitive__namespace.Root, _extends._({
        ref: ref,
        className: drawer12s.cn('bg-secondary relative h-4 w-full overflow-hidden rounded-full', className)
    }, props, {
        children: /*#__PURE__*/ jsxRuntime.jsx(ProgressPrimitive__namespace.Indicator, {
            className: "bg-primary h-full w-full flex-1 transition-all",
            style: {
                transform: `translateX(-${100 - (value || 0)}%)`
            }
        })
    }));
});
Progress.displayName = ProgressPrimitive__namespace.Root.displayName;

exports.Progress = Progress;
